### ISSUE TITLE
Crop.xml converter

### ISSUE THREAD

**Issue by HamishBrownPFR** [APSIMInitiative/ApsimX#2942]
It would be good to have a tool that we can run locally, when refactoring crop model code, that runs each crop.xml file through the converter and saves it so these changes can be comitted.  Currently we have to copy each model into a replacement node in a .apsimx file, save it, open it so the converter is applied to the xml and then copy this back to the crop.xml in the source code.

**Comment by hol430** [comment]
Would this be a standalone tool, or a feature built into Apsim? I'm guessing most regular users won't want or need something like this.

**Comment by HamishBrownPFR** [comment]
This might not be straight forward.

I have just gone through doing this manually and the converter put a whole lot of stuff in that doesn't need to be there such as <enabled> tags.  This is a problem because it obscures the intended changes in version control.  I am now repeating the process and changing the version number in the .apsimx file I am using to be only one prior so only the intended changes are applied.

Conclusion.   Any converter tool would need to specify which version changes to apply to avoid putting unnecessary xml into the crop files

**Comment by HamishBrownPFR** [comment]
After repeating the exercise I can further conclude, it is not the converter but the GUI that adds all the <activated> tags and reorders things as well.  This provides greater justification for producing a converter tool for crop xml files and the current procedure, putting them throuh the GUI to convert them, creates lots of cruft.

**Comment by HamishBrownPFR** [comment]
@hol430 this would be a standalone tool that would only be used by people who are making changes to plant code

**Comment by HamishBrownPFR** [comment]
I wonder if there are any parallels here with #2486.  A little GUI tool that opened a crop.xml and applied a converter on opening would be awsome

**Comment by hol430** [comment]
Yeah it would be handy. We could have this accessible through Apsim, but only available to people building from source, by using the `#if DEBUG........#endif` syntax.

**Comment by hol430** [comment]
So this tool just needs to open an XML file, run the converter to get this file up to the latest version, and then save the XML file?

**Comment by hol353** [comment]
- The crop XML files don't have a version number in them. Would have to ask user what version to convert from / to. 
- The crop XML files don't exist in the released version of APSIM. Could use that fact to determine whether the tool is made available to user as a menu item in GUI.
- Rather than a separate tool, could be integrated into the GUI as a button on the home screen that is only available if crop XML files exist. The GUI wouldn't load the file like a normal .apsimx file so the 'cruft' shouldn't get added.

**Comment by hol430** [comment]
Will the user ever need to take an xml file from a new version to an older one?

**Comment by hol430** [comment]
*Should* the XML files have a version number in them?

**Comment by hol353** [comment]
Yes, we should probably add a version number at the top, the same as we do for .apsimx files. If we do this then perhaps we don't need to ask for a from/to version number.

**Comment by hol430** [comment]
I now have a functional prototype of this... it's certainly modifying the xml file, but is there an easy way to know whether the changes are correct?

**Comment by hol430** [comment]
The converter's UI is now pretty much working, but now I'm encountering some difficulties with the converter itself. I've run the converter on all the files under Models\Resources, but it's throwing an exception while upgrading Potato.xml to version 32. 

This particular converter needs to operate on all factor names of all series in the file. For whatever reason, it assumes that some ancestor of a Series will be either a Simulation, Zone, Experiment, Folder, or Simulations. In the potato xml, there is a series, but there are no nodes of any of these types so this assumption falls apart.

Is it normal/acceptable to put a series node in a crop xml file?

**Comment by hol353** [comment]
There shouldn't be any <Graph> nodes in a crop xml file. They are a left over from a bug that used to exist. Delete the one in potato.

**Comment by hol353** [comment]
The message above should say "There shouldn't be any graph nodes..."

**Comment by hol430** [comment]
Cheers, will do.

**Comment by hol430** [comment]
Sorry for the double post, but I have another question. Currently, the crop xml files don't have any version associated with them, so the converter assumes they're version 0. Is this a reasonable assumption?

**Comment by hol353** [comment]
Nope not a good assumption. Better to assume they all have a version number. You'll have to manually go and add a version number to each crop.xml file which is one less than the current version.
